## Citations

1. Marçais, G. *et al.* [MUMmer4: A fast and versatile genome alignment system](https://doi.org/10.1371/journal.pcbi.1005944). *PLoS computational biology* **14**, e1005944 (2018).
2. Travers, K. J., Chin, C.-S., Rank, D. R., Eid, J. S. & Turner, S. W. [A flexible and efficient template format for circular consensus sequencing and SNP detection](https://doi.org/10.1093/nar/gkq543). *Nucleic Acids Research* **38**, e159–e159 (2010).
3. Li, H. [Minimap2: Pairwise alignment for nucleotide sequences](https://doi.org/10.1093/bioinformatics/bty191). *Bioinformatics* **34**, 3094–3100 (2018).
4. Quinlan, A. R. & Hall, I. M. [BEDTools: A flexible suite of utilities for comparing genomic features](https://doi.org/10.1093/bioinformatics/btq033). *Bioinformatics* **26**, 841–842 (2010).
5. Krueger, F., James, F., Ewels, P., Afyounian, E. & Schuster-Boeckler, B. FelixKrueger/TrimGalore: V0.6.7 - DOI via Zenodo. (2021) doi:[10.5281/ZENODO.5127899](https://doi.org/10.5281/ZENODO.5127899).
6. Thomer, A. K., Twidale, M. B., Guo, J. & Yoder, M. J. Picard Tools. in *Conference on Human Factors in Computing Systems - Proceedings* (2016).
7. Li, H. *et al.* [The Sequence Alignment/Map format and SAMtools](https://doi.org/10.1093/bioinformatics/btp352). *Bioinformatics* **25**, 2078–2079 (2009).
8. Benjamin, D. *et al.* Calling Somatic SNVs and Indels with Mutect2. (2019) doi:[10.1101/861054](https://doi.org/10.1101/861054).
9. Garrison, E. & Marth, G. Haplotype-based variant detection from short-read sequencing. (2012).

